walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,97 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public static final int DAY_START = TimeRange.START_OF_DAY;
+    public static final int DAY_END = TimeRange.END_OF_DAY;
+    public static final TimeRange WHOLE_DAY = TimeRange.WHOLE_DAY;
+
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        long meetingDuration = request.getDuration();
+        // Required attendees for the request meeting.
+        Collection<String> requiredAttendees = request.getAttendees(); 
+        // All attendees (required and optional) for the request meeting.
+        Collection<String> allAttendees = new ArrayList<>(requiredAttendees);
+        allAttendees.addAll(request.getOptionalAttendees());
+
+        // If the meeting duration is invalid, return an empty array list as 
+        // the meetingTimes collection.
+        if (meetingDuration > WHOLE_DAY.duration() || meetingDuration == 0) {
+            return Arrays.asList();
+        }
+
+        // Find available TimeRange for both required and optional attendees.
+        Collection<TimeRange> allMeetingTimes = 
+            findMeetingTimes(events, allAttendees, meetingDuration);
+
+        // If there is meeting time for all, or there is no required attendee, 
+        // return meetingTimes for all attendees.
+        if (allMeetingTimes.size() > 0 || requiredAttendees.size() == 0) {
+            return allMeetingTimes;
+        }
+        // If not, return available TimeRange for required attendees.
+        return findMeetingTimes(
+            events, requiredAttendees, meetingDuration);
+    }
+
+    /**
+      * Find all times that the meeting could happen, given attendees and duration. 
+      * Pointer rangeStart is pointing at the start of the current available 
+      * TimeRange. If the current available TimeRange has longer duration than the 
+      * requested meeting duration, add it to available meetingTimes.
```

Comment:
It's unclear what the `rangeStart` is used for. The second sentence also seem to lack context (how is the 'current available TimeRange' computed?)

Reply:
The comment is updated!